icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1268,12 +1268,31 @@ public String replace() {
                 // Nothing changed in this iteration, break out the loop
                 break;
             }  // while(1)
-            if (changed) {
-                String result =  lscvToID(language, script, region,
+            if (extensions == null && !changed) {
+                return null;
+            }
+            String result =  lscvToID(language, script, region,
                     ((variants == null) ? "" : Utility.joinStrings("_", variants)));
-                if (extensions != null) {
-                    result += extensions;
+            if (extensions != null) {
+                boolean keywordChanged = false;
+                ULocale temp = new ULocale(result + extensions);
+                Iterator<String> keywords = temp.getKeywords();
+                while (keywords != null && keywords.hasNext()) {
+                    String key = keywords.next();
+                    String value = temp.getKeywordValue(key);
```

Comment:
As in C++, I propose fetching the value only if the key is one of the two relevant ones.